games/xworld3d/x3scene.cpp

```diff
@@ -0,0 +1,160 @@
+    }
+    CHECK_GT(agents_.size(), 0) << "There should be at least one agent.";
+}
+
+void X3Scene::add_item(const X3ItemInfo& info) {
+    if (info.type == AGENT) {
+        CHECK(agents_.find(info.name) == agents_.end())
+                << "An agent with name " << info.name << " exists.";
+        auto agent = std::make_shared<X3Agent>(info, world_,
+                                               FLAGS_x3_speed_norm,
+                                               FLAGS_x3_orientation_bins,
+                                               FLAGS_x3_reaching_distance);
+        agents_[info.name] = agent;
+    } else if (info.type == GOAL) {
+        CHECK(goals_.find(info.name) == goals_.end())
+                << "A goal with name " << info.name << " exists.";
+        auto goal = std::make_shared<X3Goal>(info, world_);
+        goals_[info.name] = goal;
+    } else if (info.type == BLOCK) {
+        CHECK(blocks_.find(info.name) == blocks_.end())
+                << "An item with name " << info.name << " exists.";
+        auto block = std::make_shared<X3Block>(info, world_);
+        blocks_[info.name] = block;
+    } else {
+        LOG(ERROR) << "unknown item type: " << info.type;
+    }
+}
+
+void X3Scene::apply_action(const X3AgentPtr& agent, int a,
+                           float& reward, int& event) {
+    reward = STEP_PENALTY;
+    event = X3Event::NOTHING;
+
+    switch (a) {
+        case X3NavAction::MOVE_FORWARD:
+            agent->move_forward();
+            break;
+        case X3NavAction::TURN_LEFT:
+            agent->turn_left();
+            break;
+        case X3NavAction::TURN_RIGHT:
+            agent->turn_right();
+            break;
+        case X3NavAction::JUMP:
+            agent->jump();
+            break;
+        case X3NavAction::COLLECT:
+            {
+                X3Goal* goal = collect_goal(agent);
+                if (goal) {
+                    remove_goal(goal);
+                    reward = COLLECT_GOAL;
+                    event = X3Event::CORRECT_GOAL;
+                } else {
+                    reward = WRONG_OP;
+                    event = X3Event::WRONG_ACTION;
+                }
+            }
+            break;
+        default:
+            LOG(ERROR) << "unknown action id: " << a;
+            event = X3Event::WRONG_ACTION;
+    }
+}
+
+roboschool::RenderResult X3Scene::render(std::string agent_name, bool debug) {
+    auto it = agents_.find(agent_name);
+    CHECK(it != agents_.end())
+            << "agent " << agent_name << " not found.";
+    return camera_->render(it->second.get(), debug);
+}
+
+void X3Scene::remove_goal(X3Goal* goal) {
+    auto it = goals_.find(goal->name());
+    if (it != goals_.end()) {
+        auto& g = it->second;
+        world_.remove_object(g->object());
+        g->destroy();
+        goals_.erase(it);
+    }
+}
+
+X3Goal* X3Scene::collect_goal(const X3AgentPtr& agent) {
+    X3Goal* goal = NULL;
+    float best_score = 0.5;
```

Comment:
Explain why the best score is 0.5

Reply:
if no score is higher than 0.5, we consider the collecting action failed